### ISSUE TITLE
Feature request: less-strict prefer-const

### ISSUE THREAD

**Issue by NfNitLoop** [denoland/deno_lint#1125]
Problem
---------

There are many use cases where `prefer-const` is a bit too strict for my taste. (And I expect/hope many others.)

Consider this:

```typescript
myFn() {
    let obj = await otherFn()
    obj.value += 1
    return obj
}
```

Currently, `prefer-const` will complain that I really should make that `obj` reference `const` because I don't mutate its reference.

But it doesn't matter! It's a local variable with a very small scope. And marking the variable as `const` wouldn't make it immutable, only its **reference** becomes immutable. This is a pointless distinction for almost everything but primitive data types and the odd immutable class. (But even those can be monkey-patched, even when they're `const`!)

Additionally, if I mark it as `const`, there's a chance that later I'll come back and refactor the implementation of `myFn()` to reassign `obj` and now I need to go change `obj` from a `const` to a `let`. Why? They're just slightly different ways of mutating a variable that I do not care about distinguishing between.

As a result, due to pointless linter noise, I often end up completely disabling `prefer-const` in my file or in my `deno.jsonc`. But then I lose other `prefer-const` checks that *should* be lint-checked, like top-level `let` in a module.

Proposal
----------

Make `prefer-const` *not* check function-local variables. An additional `prefer-local-const` (or `prefer-const-local`, for better sorting) could be added to check for that case if people really want it.

If the Deno team is keen on not changing the linter's default behavior, `prefer-local-const` could be enabled by default. At least that would allow me to exclude only `prefer-local-const`. But of course, my personal preference would be for it to be disabled by default.

**Comment by NfNitLoop** [comment]
> other prefer-const checks that should be lint-checked, like top-level let in a module.

Expanding a bit here:  IMO immutability is most useful when you have many scopes of code that will share the same variable. If you're not going to mutate the value, it's handy to enforce that in one place so that you don't have to go check all code that can access the variable to see how it's used. But, when you're dealing with local methods that overhead becomes too much. Especially when it's on objects that are themselves mutable (so the distinction is meaningless).

... on that topic, I was surprised to find that while there's a `prefer-const` rule, there's *not* a `prefer-readonly` lint for class properties.  It seems like for parity with `prefer-const`, `private` class properties that are never mutated could be marked as `readonly`? But again, I'd hope that that lint would only fire for things where that distinction is meaningful. (immutable primitives/types).  But we can leave that for a separate issue. 😄   I mostly add it here to show that there are more meaningful checks that `prefer-const` is currently missing, so (IMO) it shouldn't be too controversial to relax this edge case.

**Comment by magurotuna** [comment]
I think the current behavior of `prefer-const` is what many of us agree on as it acts similarly to the ESLint's prefer-const rule.
Personally, if I see some JavaScript/TypeScript code that uses `let` to declare a variable, I would expect that that variable will be re-assigned (not property overwrite) at some later point, which requires me to pay much attention to the following code in the function, because re-assignment makes the state hard to follow in general.
As you pointed out, `const` variables still allow for mutable manipulations on them, which will also require us to pay some sort of attention in order to keep track of the current state of them. However, the fact that `const` variables will never be re-assigned is still a useful guarantee that help understand what the code does by limiting the allowed operations on the variables.
So my opinion is that even if the scope is too small, using `const` is beneficial if there's no need to re-assign. Using `let` will be a strong marker for me that makes me think that "there should be code author's some intention of using `let` here. I need to pay closer attention to the state of this variable throughout this scope". I would like to avoid this additional memory (in my head) even if that makes the code a little bit redundant or hard-to-refactor.

**Comment by NfNitLoop** [comment]
I was googling "deno prefer const issue" to find and link to this issue again, and I came across [this post] by a TC39 contributor which (despite its language and tone) may better communicate my problems with the current overzealousness of `prefer-const`.

[this post]: https://jamie.build/const

A similar issue [was considered] when adding the `var` keyword to Java for local variable type inference, and they ended up not distinguishing between mutable/immutable local references. "this would be a tradeoff of additional ceremony for additional capture of design intent".

[was considered]: https://openjdk.org/jeps/286#:~:text=this%20would%20be%20a%20tradeoff%20of%20additional%20ceremony%20for%20additional%20capture%20of%20design%20intent

And IMO, if you have the linter nagging you to change local variables to `const`, then `const` is no longer even a capture of design intent, and is instead a capture of *the project's current linter rules*.  If I want to *actually* capture design intent, I need to either litter my code with `// deno-lint-ignore prefer-const`, or continue to drop `deno.jsonc` files in everything to disable this rule project-wide, which is [my current practice].

[my current practice]: https://github.com/NfNitLoop/better-iterators/blob/a4b05812730c5e813814e6c216c15b0978ba892c/deno.jsonc#L5-L9

Would there be an objection to a PR that implements this change if I get annoyed enough to do it myself one weekend? 😊

**Comment by magurotuna** [comment]
While I really understand the feeling on the badly designed `const` and `let` in JavaScript (as a Rust developer😃), I have got used to this quirk so I wouldn't like to make some changes to `prefer-const` rule. (Also wouldn't like to separate rules to two, strict one and relaxed one).
Still in my opinion the use of `let` communicates the fact that re-assignment will occur, which likely helps us reading through complicated code. I think so probably because it's said that declarative (as opposed to procedural) style of programming is considered good in modern era. The possibility that re-assignment will happen makes me think that there should be some good reason for the code author to need to use `let` to achieve something.

**Comment by lucacasonato** [comment]
I'll close this - the rule is working as intended. We could consider adding a second "prefer-const-primitives" rule.
